src/lib/tab_groups.ts

```diff
@@ -0,0 +1,330 @@
+    const tabs = await tgroupRun(name)
+    tabs.sort((a, b) => b.lastAccessed - a .lastAccessed)
+    if (previous) {
+        return tabs[1].id
+    } else {
+        return tabs[0].id
+    }
+}
+
+/**
+ * Clear stored information for a tab group.
+ *
+ * @param name The name of the tab group.
+ *
+ */
+export async function tgroupClearOldInfo(oldName: string, newName?: string) {
+    const promises = []
+    const groups = await tgroups()
+    groups.delete(oldName)
+    if (newName) {
+        groups.add(newName)
+    }
+    promises.push(setTgroups(groups))
+
+    if (newName) {
+        promises.push(setWindowTgroup(newName))
+        promises.push(
+            tgroupRun(oldName).then(async tabs => {
+                await setTabTgroup(newName, tabs.map(tab => tab.id))
+            })
+        )
+    }
+    return Promise.all(promises)
+}
+
+/**
+ * Activate the previously active tab in a tab group.
+ *
+ * @param name The name of the tab group to switch to.
+ *
+ */
+export async function tgroupActivate(name: string) {
+    const lastActiveTabId = await tgroupLastTabId(name)
+    // this will trigger tgroupHandleTabActivated
+    return browserBg.tabs.update(lastActiveTabId, { active: true })
+}
+
+/**
+ * Activate the last active tab in the last active tab group.
+ *
+ */
+export async function tgroupActivateLast() {
+    const otherGroupsTabs = await tgroupRun(await windowTgroup() as string, true)
+    otherGroupsTabs.sort((a, b) => (b.lastAccessed - a .lastAccessed))
+    return browserBg.tabs.update(otherGroupsTabs[0].id, { active: true })
+}
+
+/**
+ * Clear all stored tab group information.
+ *
+ */
+export async function clearAllTgroupInfo() {
+    return Promise.all([
+        clearTgroups(),
+        clearWindowTgroup(),
+        browser.tabs.query({currentWindow: true}).then(async tabs => {
+            const ids = tabs.map(tab => tab.id)
+            await browser.tabs.show(ids)
+            return clearTabTgroup(ids)
+        })
+    ])
+}
+
+/**
+ * Set the tab's tab group to the window's active tab group if there is one.
+ *
+ * Do nothing if the tab is already associated with a tab group.
+ *
+ * @param tab The tab that was just created.
+ *
+ */
+export async function tgroupHandleTabCreated(tab: browser.tabs.Tab) {
+    const windowGroup = await windowTgroup() as string
+
+    if (windowGroup) {
+        const tabGroup = await tabTgroup(tab.id)
+        if (!tabGroup) {
+            return setTabTgroup(windowGroup, tab.id)
+        }
+    }
+}
+
+/**
+ * Set the tab's tab group to the window's active tab group if there is one.
+ *
+ * @param tabId The id of the tab that was just attached to this window.
+ *
+ */
+export async function tgroupHandleTabAttached(tabId: number, _attachInfo: any) {
+    // NOTE this doesn't need to worry about a tab on another window previously
+    // being pinned because tabs become unpinned when you move them between
+    // windows
+    const windowGroup = await windowTgroup() as string
```

Comment:
```suggestion
    const windowGroup = await windowTgroup()
```
If you add the cas to windowTgroup.